### PR TITLE
team-level std algos: part 10

### PR DESCRIPTION
**Pull request by fnrizzi**
Co-authored-by: Cezary Skrzyński
Co-authored-by: Jakub Strzebonski

## Content 

Team-level implementation and tests for:

- `Kokkos_IsSortedUntil.hpp` 
- `Kokkos_IsSorted.hpp` 
- `Kokkos_IsPartitioned.hpp` 
- `Kokkos_PartitionCopy.hpp` 
- `Kokkos_PartitionPoint.hpp`

## Technical comments

- the change `class` -> `typename` and the enable if defaulted in the template parameters comes from [this request](https://github.com/kokkos/kokkos/pull/6200#issuecomment-1626406816)

- the code duplication is intentional and agreed upon after the long discussion in [here](https://github.com/kokkos/kokkos/pull/6200)

- tests are run for both contiguous views and strided views

## Other comments

- all team-level PRs are tracked in: https://github.com/kokkos/kokkos/issues/6352

- this comes from breaking apart https://github.com/kokkos/kokkos/pull/5150

## Possible conflicts

Implementation-wise, this has not conflict with other team-level PRs so it could be merged independently. 
However, a trivial conflict in the CMakeLists inside the unit test occurs if another of the team-level PRs is merged before this because each team-level PR adds test using a specific "label". Resolving this is trivial. 

## Associated wiki PRs

- `Kokkos_IsSortedUntil.hpp`: see [wiki PR](https://github.com/kokkos/kokkos-core-wiki/pull/436)
- `Kokkos_IsSorted.hpp`, `Kokkos_IsPartitioned.hpp`, `Kokkos_PartitionCopy.hpp`: see [wiki PR](https://github.com/kokkos/kokkos-core-wiki/pull/403) 
- `Kokkos_PartitionPoint.hpp`: see [wiki PR](https://github.com/kokkos/kokkos-core-wiki/pull/440) 

IMPORTANT: wiki PRs are not complete yet but 99% there 

The issue tracking all *wiki PRs* is [here](https://github.com/kokkos/kokkos-core-wiki/issues/379)

### CODE REVIEW

**Comment by fnrizzi** [comment]
Retest this please

**Comment by fnrizzi** [comment]
retest this please

**Comment by fnrizzi** [comment]
openmptarget failure unrelated

File: algorithms/src/std_algorithms/impl/Kokkos_PartitionCopy.hpp [hunk 1]
```diff
@@ -126,6 +125,55 @@ partition_copy_impl(const std::string& label, const ExecutionSpace& ex,
...
+
+  if (from_first == from_last) {
+    return {to_first_true, to_first_false};
+  }
+
+  const std::size_t num_elements =
+      Kokkos::Experimental::distance(from_first, from_last);
+
+  // FIXME: there is no parallel_scan overload that accepts TeamThreadRange and
+  // return_value, so temporarily serial implementation is used
```
Comment (masterleinad) [inline algorithms/src/std_algorithms/impl/Kokkos_PartitionCopy.hpp:155]: This is changing with recent pull requests, though.

**Review by masterleinad — APPROVED** [review]
Fine with me.

File: algorithms/src/std_algorithms/impl/Kokkos_PartitionCopy.hpp (hunk 1 above)
Comment (fnrizzi) [reply algorithms/src/std_algorithms/impl/Kokkos_PartitionCopy.hpp:155]: yes, i am going to fix all of this *after* all the team PRs are merged in a follow up.

**Review by crtrott — APPROVED** [review]

**Pull request merged**